File: admin-backend/admin-auth/src/main/java/io/github/faustofan/admin/auth/domain/service/UserDetailsServiceImpl.java
package io.github.faustofan.admin.auth.domain.service;

import java.util.Objects;
import java.util.stream.Collectors;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.cache.annotation.CacheConfig;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

import io.github.faustofan.admin.auth.domain.model.LoginUser;
import io.github.faustofan.admin.shared.cache.constants.CacheKeys;
import io.github.faustofan.admin.shared.common.context.AppContextHolder;
import io.github.faustofan.admin.shared.common.exception.BizException;
import io.github.faustofan.admin.shared.common.exception.UserException;
import io.github.faustofan.admin.shared.common.exception.errcode.BizErrorCode;
import io.github.faustofan.admin.shared.common.exception.errcode.UserErrorCode;
import io.github.faustofan.admin.shared.distributed.constants.RedisKeyRegistry;
import io.github.faustofan.admin.shared.distributed.core.RedisUtil;
import io.github.faustofan.admin.system.domain.enums.UserStatus;
import io.github.faustofan.admin.system.dto.SysUserLoginView;
import io.github.faustofan.admin.system.infrastructure.reponsitory.SysUserRepository;
import jakarta.servlet.http.HttpServletRequest;

/**
 * 用户详情服务
 * 实现 Spring Security 的 UserDetailsService 接口
 */
@Service
@CacheConfig(cacheNames = CacheKeys.CACHE_AUTH_USER)
public class UserDetailsServiceImpl implements UserDetailsService {

    private final Logger logger = LoggerFactory.getLogger(this.getClass());

    private final SysUserRepository userRepository;

    private final RedisUtil redisUtil;

    public UserDetailsServiceImpl(SysUserRepository userRepository, RedisUtil redisUtil) {
        this.userRepository = userRepository;
        this.redisUtil = redisUtil;
    }

    /**
     * 根据用户名加载用户信息
     * Spring Security 标准接口，多租户场景下请使用 loadUserByUsernameAndTenant
     */
    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        throw new UsernameNotFoundException("请使用 loadUserByUsernameAndTenant 方法");
    }

    /**
     * 场景 A: 登录时加载
     * 缓存 Key 格式: name:{tenantId}:{username}
     * sync = true: 防止高并发登录击穿 DB
     */
    @Cacheable(key = CacheKeys.KEY_NAME_SPEL + "+ #tenantId + ':' + #username", sync = true)
    public LoginUser loadUserByUsernameAndTenant(String username, Long tenantId) {
        LoginUser user = loadUserFromDatabase(username, tenantId);
        if (user == null) {
            throw new BizException(BizErrorCode.USER_NOT_EXIST_OR_DISABLED);
        }
        return user;
    }

    /**
     * 场景 B: 刷新 Token / 获取当前用户信息时加载
     * 缓存 Key 格式: id:{userId}
     * 注意：这里和上面是不同的 Key，缓存了两份数据。
     * 权衡：冗余存储换取 O(1) 的读取速度，更新时需要同时清除两份。
     */
    @Cacheable(key = CacheKeys.KEY_ID_SPEL + "+ #userId", sync = true)
    public LoginUser loadUserById(Long userId) {
        var userEntity = userRepository.findByIdWithRoles(userId);
        if (userEntity == null) {
            throw new BizException(BizErrorCode.USER_NOT_EXIST);
        }
        LoginUser user = loadUserFromDatabase(userEntity.getUsername(), userEntity.getTenantId());
        if (user == null) {
            throw new BizException(BizErrorCode.USER_NOT_EXIST);
        }
        return user;
    }

    /**
     * 检查 Token 是否已登出
     * @param token JWT Token
     * @return      true 如果已登出，false 否则 
     */
    public boolean isLogOut(String token) {
        return redisUtil.isMemberOfSet(RedisKeyRegistry.SEC_BLACKLIST, "TOKENS", token);
    }

    /**
     * 从数据库加载用户信息
     */
    private LoginUser loadUserFromDatabase(String username, Long tenantId) {
        var user = userRepository.findByUsernameWithRoles(username, tenantId);
        if (user == null) {
            return null;
        }
        if (user.getStatus() != UserStatus.ACTIVE) {
            throw new BizException(BizErrorCode.USER_DISABLED);
        }

        var roles = user.getRoles();
        boolean isSuperAdmin = user.isSuperAdmin();

        var roleMaps = roles.stream()
                .collect(Collectors.toUnmodifiableMap(
                        SysUserLoginView.TargetOf_roles::getId,
                        SysUserLoginView.TargetOf_roles::getName
                ));

        var permissions = roles.stream()
                .flatMap(role -> role.getMenus().stream())
                .map(SysUserLoginView.TargetOf_roles.TargetOf_menus::getPermCode)
                .filter(Objects::nonNull)
                .collect(java.util.stream.Collectors.toUnmodifiableSet());

        var policyContents = roles.stream()
                .flatMap(role -> role.getPolicies().stream())
                .map(SysUserLoginView.TargetOf_roles.TargetOf_policies::getPolicyContent)
                .collect(Collectors.toUnmodifiableSet());

        return new LoginUser(
                user.getId(),
                user.getTenantId(),
                user.getOrg().getId(),
                user.getOrg().getName(),
                user.getUsername(),
                user.getPassword(),
                user.getNickname(),
                roleMaps,
                permissions,
                policyContents,
                isSuperAdmin,
                user.getStatus() == UserStatus.ACTIVE,
                true,
                true,
                true);
    }
}

File: src/main/kotlin/io/github/faustofan/admin/auth/domain/service/UserDetailsServiceImpl.kt
package io.github.faustofan.admin.auth.domain.service

import io.github.faustofan.admin.auth.domain.model.LoginUser
import io.github.faustofan.admin.shared.application.context.RoleDataScopeInfo
import io.github.faustofan.admin.shared.infrastructure.exception.BizException
import io.github.faustofan.admin.system.domain.model.UserStatus
import io.github.faustofan.admin.system.infrastructure.persistence.SysUserRepository
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service

/**
 * 用户详情服务
 *
 * 实现 Spring Security 的 UserDetailsService 接口
 * 用于登录时加载用户信息
 *
 * 注意：缓存由 LoginUserCacheService 统一处理，本类只负责业务逻辑
 */
@Service
class UserDetailsServiceImpl(
    private val userRepository: SysUserRepository,
    private val loginUserCacheService: LoginUserCacheService
) : UserDetailsService {

    /**
     * 根据用户名加载用户信息
     * Spring Security 标准接口，多租户场景下请使用 loadUserByUsernameAndTenant
     */
    override fun loadUserByUsername(username: String): UserDetails {
        throw UsernameNotFoundException("请使用 loadUserByUsernameAndTenant 方法")
    }

    /**
     * 根据用户名和租户ID加载用户信息
     */
    fun loadUserByUsernameAndTenant(username: String, tenantId: Long): LoginUser {
        return loginUserCacheService.getByUsernameAndTenant(username, tenantId) {
            loadUserFromDatabase(username, tenantId)
        } ?: throw BizException(message = "用户不存在或已被禁用")
    }

    /**
     * 根据用户 ID 加载用户信息（用于 Token 刷新）
     */
    fun loadUserById(userId: Long): LoginUser {
        return loginUserCacheService.getByUserId(userId) {
            val user = userRepository.findByIdWithRoles(userId)
                ?: throw BizException(message = "用户不存在")
            loadUserFromDatabase(user.username, user.tenantId)
        } ?: throw BizException(message = "用户不存在")
    }

    /**
     * 从数据库加载用户信息
     */
    private fun loadUserFromDatabase(username: String, tenantId: Long): LoginUser? {
        val user = userRepository.findByUsernameWithRoles(username, tenantId)
            ?: return null

        if (user.status != UserStatus.ACTIVE) {
            throw BizException(message = "用户已被禁用")
        }

        val permissions = mutableSetOf<String>()
        val roleDataScopes = mutableListOf<RoleDataScopeInfo>()
        var isSuperAdmin = false

        for (role in user.roles) {
            if (role.code == "super_admin") {
                isSuperAdmin = true
            }

            for (menu in role.menus) {
                menu.permission?.let { permissions.add(it) }
            }

            for (permission in role.permissions) {
                permissions.add(permission.code)
            }

            roleDataScopes.add(
                RoleDataScopeInfo(
                    dataScope = role.dataScope,
                    specificOrgIds = role.specificOrgs.map { it.id }
                )
            )
        }

        return LoginUser(
            userId = user.id,
            tenantId = user.tenantId,
            orgId = user.org?.id,
            username = user.username,
            password = user.passwordHash,
            nickname = user.nickname,
            permissions = permissions,
            roleDataScopes = roleDataScopes,
            isSuperAdmin = isSuperAdmin,
            enabled = user.status == UserStatus.ACTIVE
        )
    }
}

